### DISCUSSION TITLE
Can't add a seamless DJ script

### DISCUSSION THREAD

**Discussion by Bloofield** [savonet/liquidsoap#4066] [Q&A]
I'm trying to make it so that when a DJ connects, the next song doesn't automatically start playing,

I'm stuck within the bounds of using Azuracast.

Here's my code, it keeps complaining about the thread.run, what am I doing wrong?

```
# Attempt to switch from AutoDJ to live
def check_to_live_and_transition(old, new) =
    log("Live is enabled, waiting for 15 seconds to see if to_live becomes true.")
    stop_loop = false

    # Define the function to be run in a separate thread

    for i = 1 to 15 do
        if not stop_loop then          
            if to_live() then
                stop_loop = true
            else
                log("Waiting... #{i} seconds")
                thread.pause(1.0)
            end
        end
    end
    if stop_loop then
        log("to_live became true within 15 seconds, transitioning to live.")
        sequence([
            fade.out(duration=settings.autocue.cue_file.fade_out(), old.source),
            fade.in(duration=settings.autocue.cue_file.fade_in(), new.source)
        ])
    else
        log("to_live did not become true within 15 seconds, fading into the next song.")
        perform_crossfade(old, new)
    end
end

# Main function for live-aware crossfade
def live_aware_crossfade(old, new) =
    if live_enabled() then
        thread.run(false, 0., null, check_to_live_and_transition(old, new))
    else
        perform_crossfade(old, new)
    end
end
```

**Comment by vitoyucepi** [comment]
Hi @Bloofield,
It's quite challenging for me to comprehend the desired behavior, but here are some thoughts.

The `thread.run` is supposed to take a function as a first positional argument, and the rest will be named arguments.
```ruby
def f()
  # do something
  ()
end

thread.run(f)
```

The named arguments are
- `fast` - if the function call can be executed without blocking others. Slow operations should run in `fast = false` mode.
- `delay` - is a delay before a function is called for the first time.
- `every` - is a parameter that could be either a constant value or a function that controls time before the next function call.

The `thread.run` function could not be used in this context because it returns nothing
```ruby
def live_aware_crossfade(old, new) =
    if live_enabled() then
        thread.run({check_to_live_and_transition(old, new)})
        # implicit crossfade should be placed here
    else
        perform_crossfade(old, new)
    end
end
```

The purpose of `check_to_live_and_transition` is not clear
```ruby
def check_to_live_and_transition(old, new)
  stop_loop = false
  for i = 1 to 15 do
    if not stop_loop then          
      if to_live() then
        stop_loop = true
      else
        log("Waiting... #{i} seconds")
        thread.pause(1.0)
      end
    end
  end

  # do something
  ()
end
```

You can't assign the value from outside of the scope from inside without using `ref`.
```ruby
stop_loop = ref(false)
if not stop_loop() then
  stop_loop := true
end
```

**Reply by Bloofield** [reply]
Thanks for this, I'm still not sure what to do here though.

For context, I'm trying to make it so that this happens:

1. When nearing the end of a song, check if a DJ is currently connected and ready to start their own music (i.e. take over from the AutoDJ) 
2. When we've confirmed a DJ is connected, we'll give them about 15s of radio silence to start playing
3. If the DJ does not start playing within the 15s window, we'll go back to the AutoDJ

Initially I had it 'working', but I was using thread.pause but didn't realise that it's not for use in the main stream.

Here's the original code generated by Azuracast

```
# Fading/crossing/segueing
def live_aware_crossfade(old, new) =
    if to_live() then
        # If going to the live show, play a simple sequence
        # fade out AutoDJ, do (almost) not fade in streamer
        sequence([
          fade.out(duration=settings.autocue.cue_file.fade_out(), old.source),
          fade.in(duration=settings.autocue.cue_file.fade_in(), new.source)
        ])
    else
        # Otherwise, use a beautiful add
        add(normalize=false, [
            fade.in(
                initial_metadata=new.metadata,
                duration=settings.autocue.cue_file.fade_in(),
                new.source
            ),
            fade.out(
                initial_metadata=old.metadata,
                duration=settings.autocue.cue_file.fade_out(),
                old.source
            )
        ])
    end
end

radio = cross(
    duration=settings.autocue.cue_file.fade_out(),
    live_aware_crossfade,
    radio
)
```

I was overwriting it by calling `cross` & `live_aware_crossfade` a bit earlier in the code (as we can't change Azuracast's generated code)

Thank you so much for your help
